packages/css/src/plugins/atomicify-rules.tsx

```diff
@@ -0,0 +1,215 @@
+  // Because of that we trim whitespace as well as removing any top level nesting "&".
+  const trimmed = selector.trim();
+
+  switch (trimmed.charAt(0)) {
+    case ':':
+      // Dangling pseudo - return immedately!
+      return trimmed;
+
+    case '&':
+      // Dangling nesting - replace it and return!
+      return trimmed.replace('&', '');
+
+    default:
+      // Must be a nested selector - add a space before it!
+      return ` ${trimmed}`;
+  }
+};
+
+/**
+ * Replaces all instances of a nesting operator `&` with the parent class name.
+ *
+ * @param selector
+ * @param parentClassName
+ */
+const replaceNestingSelector = (selector: string, parentClassName: string) => {
+  return selector.replace(/ &/g, ` .${parentClassName}`);
+};
+
+/**
+ * Builds an atomic rule selector.
+ *
+ * @param node
+ */
+const buildAtomicSelector = (node: Declaration, opts: AtomicifyOpts) => {
+  const selectors: string[] = [];
+
+  (opts.selectors || ['']).forEach((selector) => {
+    const normalizedSelector = normalizeSelector(selector);
+    const className = atomicClassName(node.prop, node.value, {
+      ...opts,
+      selectors: [normalizedSelector],
+    });
+    const replacedSelector = replaceNestingSelector(normalizedSelector, className);
+
+    selectors.push(`.${className}${replacedSelector}`);
+
+    if (opts.callback) {
+      opts.callback(className);
+    }
+  });
+
+  return selectors.join(', ');
+};
+
+/**
+ * Transforms a declaration into an atomic rule.
+ *
+ * @param node
+ * @param opts
+ */
+const atomicifyDecl = (node: Declaration, opts: AtomicifyOpts) => {
+  const selector = buildAtomicSelector(node, opts);
+  const newDecl = decl({ prop: node.prop, value: node.value });
+  const newRule = rule({ selector, nodes: [newDecl] });
+
+  // We need to link the new node to a parent else autoprefixer blows up.
+  newDecl.parent = newRule;
+  newDecl.raws.before = '';
+  newRule.parent = opts.parentNode!;
+  newRule.raws.before = '';
+
+  return newRule;
+};
+
+/**
+ * Transforms a rule into atomic rules.
+ *
+ * @param node
+ * @param opts
+ */
+const atomicifyRule = (node: Rule, opts: AtomicifyOpts) => {
+  if (!node.nodes) {
+    return [];
+  }
+
+  return node.nodes.map((childNode) => {
+    if (childNode.type !== 'decl') {
+      throw childNode.error(`Nested ${childNode.type}s are not allowed.`);
+    }
+
+    return atomicifyDecl(childNode, {
+      ...opts,
+      selectors: node.selectors,
+    });
+  });
+};
+
+/**
+ * Transforms an atrule into atomic rules.
+ *
+ * @param node
+ * @param opts
+ */
+const atomicifyAtRule = (node: AtRule, opts: AtomicifyOpts): AtRule => {
+  const children: Node[] = [];
+  const newNode = node.clone({ nodes: children });
+  const atRuleLabel = `${opts.atRule || ''}${node.name}${node.params}`;
+  const atRuleOpts = {
+    ...opts,
+    parentNode: newNode,
+    atRule: atRuleLabel,
+  };
+
+  newNode.parent = opts.parentNode!;
+
+  node.each((childNode) => {
+    switch (childNode.type) {
```

Comment:
nit: This logic seems repeated on [Line 197](https://github.com/atlassian-labs/compiled/pull/299/files#diff-f3a5a559c0a58e66119a601b2e03c384R197). Not sure. Just 3 switch cases and loop are same 🤔

Reply:
its very similar but a little different - let's leave it for now